### PR TITLE
SIMSBIOHUB-457: Taxonomy Endpoints

### PR DESCRIPTION
**Pull request by KjartanE**
## Links to Jira Tickets

- https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-457

## Description of Changes

- Added endpoint for ITIS Solr TSN search 

## Testing Notes

- {List any relevant testing considerations, necessary pre-reqs, and areas of the app to focus on. Specifically, include anything that will help the reviewers of this PR verify the code is functioning as expected.}

### CODE REVIEW

**Review by NickPhura — DISMISSED** [review]
🌞

**Review by NickPhura — DISMISSED** [review]

File: api/src/paths/taxonomy/taxon/tsn/index.ts [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+                items: {
+                  title: 'Species',
+                  type: 'object',
+                  required: ['tsn', 'commonName', 'scientificName'],
+                  properties: {
+                    tsn: {
+                      type: 'integer'
+                    },
+                    commonName: {
+                      type: 'string'
```
Comment (curtisupshall) [inline api/src/paths/taxonomy/taxon/tsn/index.ts]: Nullable?

File: api/src/repositories/taxonomy-repository.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+   * @param {Record<any, any>} itisData
+   * @param {string} itisUpdateDate
+   * @return {*}  {Promise<TaxonRecord>}
+   * @memberof TaxonomyRepository
+   */
+  async addItisTaxonRecord(
+    itisTsn: number,
+    itisScientificName: string,
+    commonName: string | null,
+    itisData: Record<any, any>,
```
Comment (curtisupshall) [inline api/src/repositories/taxonomy-repository.ts:63]: Do Records get treated as JSON blobs when they're inserted?

File: api/src/services/itis-service.ts [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+  /**
+   * Cleans up the ITIS search response data.
+   *
+   * @param {ItisSolrSearchResponse[]} data
+   * @memberof ItisService
+   */
+  _sanitizeItisData = (data: ItisSolrSearchResponse[]): TaxonSearchResult[] => {
+    return data.map((item: ItisSolrSearchResponse) => {
+      const commonName = (item.commonNames && item.commonNames[0].split('$')[1]) || null;
```
Comment (curtisupshall) [inline api/src/services/itis-service.ts]: Nit: use `??` coalesce

File: api/src/services/itis-service.ts [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+   */
+  async getItisSolrTermSearchUrl(searchTerms: string[]): Promise<string> {
+    const itisUrl = this._getItisSolrUrl();
+
+    if (!itisUrl) {
+      defaultLog.debug({ label: 'getItisTermSearchUrl', message: 'Environment variable ITIS_URL is not defined.' });
+      throw new Error('Failed to build ITIS query.');
+    }
+
+    // const itisSearchSpecies = this._getItisSolrSearchSpeciesQuery(searchParams);
```
Comment (curtisupshall) [inline api/src/services/itis-service.ts]: Remove

File: api/src/services/taxonomy-service.ts [hunk 5]
```diff
@@ -1,205 +1,104 @@
...
+
+    const missingTsnIds = tsnIds.filter((tsnId) => !existingTaxonRecords.find((item) => item.itis_tsn === tsnId));
+
+    if (missingTsnIds.length) {
+      // If the local database does not contain a record for all of the requested ids, search ITIS for the missing
+      // taxon records, patching the missing records in the local database in the process
+      const itisService = new ItisService();
+      const itisResponse = await itisService.searchItisByTSN(missingTsnIds);
+
+      for (const itisRecord of itisResponse) {
```
Comment (curtisupshall) [inline api/src/services/taxonomy-service.ts]: Could improve performance using `Promise.all`:
```typescript
async getTaxonByTsnIds(tsnIds: number[]): Promise<TaxonSearchResult[]> {
    // Search for taxon records in the database
    const existingTaxonRecords = await this.taxonRepository.getTaxonByTsnIds(tsnIds);
    let patchedTaxonRecords = [];

    const missingTsnIds = tsnIds.filter((tsnId) => !existingTaxonRecords.find((item) => item.itis_tsn === tsnId));

    if (missingTsnIds.length) {
      // If the local database does not contain a record for all of the requested ids, search ITIS for the missing
      // taxon records, patching the missing records in the local database in the process
      const itisService = new ItisService();
      const itisResponse = await itisService.searchItisByTSN(missingTsnIds);

      patchedTaxonRecords = await Promise.all(itisResponse.map(this.addItisTaxonRecord));
    }

    return this._sanitizeTaxonRecordsData(existingTaxonRecords.concat(patchedTaxonRecords));
```

File: api/src/repositories/taxonomy-repository.ts (hunk 2 above)
Comment (KjartanE) [reply api/src/repositories/taxonomy-repository.ts:63]: Yes they do

**Review by curtisupshall — APPROVED** [review]
Code looks good to me

**Review by NickPhura — APPROVED** [review]
🚕

**Pull request merged**